### DISCUSSION TITLE
[WASM] Memory leaks / understanding memory management

### DISCUSSION THREAD

**Discussion by DierkDroth** [unoplatform/uno#17585] [General]
Since days I'm chasing memory leaks in my UNO WASM application which don't occur in the 100% equivalent WinUI3 application. I understand that WASM still is 32bit whereas my WinUI3 app is 64bit - but that's not the point.

Before I raised more detailed questions (based on my findings) I wanted to understand:

Could I expect that if my WinUI 3 app would not show any memory leaks (e.g. confirmed by VS memory profiler) that my identical UNO WASM app should not have any memory leaks neither?
                          - or -
Are there UNO WASM scenarios which I should avoid (e.g. like at the bottom of [this page](https://platform.uno/docs/articles/uno-development/troubleshooting-memory-issues.html)), since they would exhibit memory leaks where the same scenario in WinUI 3 would not?

**Comment by jeromelaban** [comment]
Thanks for the question. Uno or WinUI are not exempt from memory leaks, but each has its own specificities and leaks can happen independent of each other, for the same code. You can track memory allocations using [the memory profiler](https://platform.uno/docs/articles/external/uno.wasm.bootstrap/doc/features-profiling.html#using-the-memory-profiler).

**Reply by DierkDroth** [reply]
Thanks for clarification @jeromelaban. That matches my research so far...

**Reply by jeromelaban** [reply]
> @jeromelaban I just found that UNO WASM has a FrameworkElement.Dispose() method, which does not exist in WinUI3.

`FrameworkElement.Dispose()` must not be used, as it can cause unrelated lifecycle issues.

> Following [these guidelines](https://learn.microsoft.com/en-us/previous-versions/xamarin/cross-platform/deploy-test/memory-perf-best-practices) this would imply that I had to unsubscribe from all events for all UI components when using UNO WASM, whereas this would not be required when using WinUI3. Is this the reality? This would render the concept of GC based memory management pretty much useless when using UNO WASM...

There are no such extremes as useless. These guidelines are for Xamarin based apps on iOS and Android, for which memory pinning in the native runtimes memory management can cause GC opacity in reference graphs. Uno for WebAssembly does not have these restrictions.

If you have memory leaks, use the tooling I posted above to find them out.

**Reply by DierkDroth** [reply]
Thanks again for clarification @jeromelaban. So far I'm not using `FrameworkElement.Dispose()` and will make sure I'm not using it in future.

After days of research - and by using 'real world' test scenario in my complex app - I already can confirm that there is a massive memory leak problem with UNO WASM: 
- the WinUI3 app reports no memory leak whatsoever (using any .NET memory profiler)
- however, with UNO WASM there are massive memory leaks (confirmed by the tooling you provided) - and the app runs into an OutOfMemory exception at some point - **on the exact same code**
- I can report that those problems can be lessened if I explicitly would detach event handlers which I really should not need to detach. Basically this is across the board:
-- all .Loading, .Loaded, .LayoutUpdated, .Tapped, .Clicked etc. standard event handlers
-- any custom event handler property
-- any other static or local event handler

I'm really quite baffled and am challenging myself if what I'm seeing is for real: there just is no way that I needed **to explicitly detach all event handlers** I could find and still would run into memory leaks (even if smaller ones, but still) 

This is UNO 5.1 (I can't really use any later version [for this reason](https://github.com/unoplatform/uno/issues/16928) - but even tried UNO 5.3 with similarly bad results)

What's going on there? Would you have any pointers?

**Reply by jeromelaban** [reply]
The best way to troubleshoot memory issues is to use the profiler tooling I mentioned above, then extract a repro.

**Reply by DierkDroth** [reply]
@jeromelaban to clarify: yes, I confirmed the described behavior by using the tooling you mentioned above.

To confirm: it's absolutely **not expected** that I needed to unsubscribe events in UNO WASM which I don't have to unsubscribe in WinUI3, correct? Or are there known cases where I have to unsubscribe events in UNO WASM (but not in WinUI3) to make sure the GC would free up the FrameworkElement/UIElement/etc.

**Reply by jeromelaban** [reply]
The best way to identify issues is to use the Xamarin profiler tool, then find exactly what is causing the leak, even though it's great you confirmed that there is a leak. With regards to events, the behavior may be different from WinUI in subtle way, even if it should not. 

If unregistering from events (it's generally a good idea, regardless) reduces the memory impact, capture this in a repro, it will be easier to troubleshoot.

**Reply by DierkDroth** [reply]
Wow @jeromelaban this worries me for several reasons:
- as mentioned my app scenarios are pretty complex and isolating them to a 'simple' repo is massive effort at least - if even possible at all
- basically you're saying that memory leaks are to be expected if not detaching **all events handlers** ('subtle differences in behavior') although that's not needed with WinUI3 at all - which would require massive codes changes throughout the app
- I have indications (but will have to look into again) that even as I detached all events handlers which I'm aware off there still are memory leaks which chased the app into an OutOfMemory scenario after a 'sufficient' number of iterations - whereas WinUI3 has no problem whatsoever

**Reply by jeromelaban** [reply]
> as mentioned my app scenarios are pretty complex and isolating them to a 'simple' repo is massive effort at least - if even possible at all

Community support can **only work by repros**, we **cannot debug your app** for you. We provide however the tools to help you narrow down further which scenarios may be causing the issue(s).

> basically you're saying that memory leaks are to be expected if not detaching all events handlers ('subtle differences in behavior') although that's not needed with WinUI3 at all - which would require massive codes changes throughout the app

This is **not** what I said.

It is common knowledge that events are leading cause of memory leaks in .NET, the Uno Platform is code base is not different from any codebase. There are certainly memory leaks we do not know about that can be caused by Uno or the .NET runtime in the context of WebAssembly, but until we know the cause, we can't tell much more.

**Comment by DierkDroth** [comment]
@jeromelaban I just found that UNO WASM has a FrameworkElement.Dispose() method, which does not exist in WinUI3.

Following [these guidelines](https://learn.microsoft.com/en-us/previous-versions/xamarin/cross-platform/deploy-test/memory-perf-best-practices) this would imply that I had to unsubscribe **from all events for all UI components** when using UNO WASM, whereas this would not be required when using WinUI3.

Is this the reality? This would render the concept of GC based memory management pretty much useless when using UNO WASM...

**Reply by francoistanguay** [reply]
Dierk, for further clarity, this is NOT true that you need to unregister from all events at all time. All Jerome said is that it could be a cause for some leaks, possibly, in some circumstances.

A discussion only brings real value if you can come up with a real example of a real leak/problem that we'd be unaware of. I will close this discussion for now and ask that you open a new one with a real example so the team has something concrete to discuss and/or to look at.

Thank you.

**Reply by DierkDroth** [reply]
Thanks for clarification @francoistanguay and @jeromelaban. That's exactly what I needed to know: if **the same code base** does not show memory leaks in WinUI3 but in UNO WASM, then this would be considered a bug (either UNO or .NET WASM or ...). Additionally **it's not required to detach event handlers** in UNO WASM if it's not required in WinUI3.

I will try to analyze further and will get back as I had more results...
